### DISCUSSION TITLE
Popularity and Connected Region color reduction algorithms

### DISCUSSION THREAD

**Discussion by Jsfasdf250** [ImageMagick/ImageMagick#5934] [Digital Image Processing]
Could these algorithms [here](https://github.com/ImageMagick/ImageMagick/issues/5927) be implemented in scripts? It's okay if they just output palettes or they output the color-reduced image (with dithering).

**Comment by snibgo** [comment]
What version of IM, on what platform?

The other thread has determined that you don't really want a script that finds the most popular/common colours.

I suggest you post some sample inputs, ideally with required outputs. Perhaps we can find what you really want. For example, perhaps we should first make transparent all pixels that are not exactly equal to its 8 neighbours, so we then examine only areas that are solid colours.

**Reply by Jsfasdf250** [reply]
I am using IM 7.1.0-57.

The solution of discarding pixels that touch pixels of different colors could fix the problem, but if the original image has much more colors than desired and the processed image (before applying the remap) has much less colors than desired, what would happen?
There is also the question of how to deal with semi-transparent colors:

- Ignore the alpha channel completely
- Use the alpha channel as some sort of weighting, while still including only fully-opaque colors in the colormap
- Treat them as colors in their own right

Since I plan on making gifs which only support boolean transparency, the second method may be the best option for me.

Here is a sample image to test the algorithm on. I was using images based on copyrighted fictional characters, so I had to make this image. The image primarily uses the 64 6-bit colors, but there are also many intermediate colors. I don't really have a specific output, but the 64 6-bit colors must all be included in the final colormap for it to be considered acceptable.

![plasma](https://user-images.githubusercontent.com/53478680/210647274-4cfa8c15-12b4-4e31-8636-35eddb8456fa.png)

**Comment by snibgo** [comment]
Your sample input has an alpha channel but all pixels are entirely opaque, so I ignore alpha.

It has 9984 unique colors.

Windows BAT script:
```
%IMG7%magick ^
  populSamp.png +write mpr:INP ^
  ( +clone -unique-colors +write info: +delete ) ^
  ( +clone ^
    -statistic StandardDeviation 3x3 ^
    -grayscale Brightness ^
    -threshold 0 ^
+write x.png ^
    -negate ^
  ) ^
  -alpha off ^
  -compose CopyOpacity -composite ^
  -background Black -alpha Background ^
+write x2.png ^
  -unique-colors ^
+write info: ^
+write map.png ^
  +write mpr:MAP +delete ^
  mpr:INP ^
  -remap mpr:MAP ^
  out.png
```
Each `+write x.png` etc is just for interest and debugging.

This makes out.png, which contains 64 colours:

![out](https://user-images.githubusercontent.com/33812368/210658062-9c04b147-09e4-4886-8d4c-93c420637f87.png)

**Reply by Jsfasdf250** [reply]
Since GIF can support up to 256 colors (255 actually because full transparency counts as a color), could you reduce the original image to exactly 255 colors (with Floyd-Steinberg dithering)?

**Comment by snibgo** [comment]
"`-colors 255`" will reduce it to a maximum of 255 colours. For this image, it is *exactly* 255.

**Comment by Jsfasdf250** [comment]
Here's another test image (reduce to 255 colors). Note that there are many fully transparent regions of different colors. ![image](https://user-images.githubusercontent.com/53478680/210669110-b32af45f-0b0a-4464-9b31-1b8bb8956744.png)
The final colormap must contain exactly 255 colors, including all of the primary colors.
